## Dropout Regularization
**Dropout regularization** is a technique used in deep learning to prevent overfitting. It works by randomly 'dropping out' (i.e. setting to zero) a number of neurons during each training iteration. This helps to reduce the complexity of the model and make the learning more robust.

This dataset describes sonar chirp returns bouncing off different services. The 60 input variables are the strength of the returns at different angles. It's a binary classification problem that requires a model to differentiate rocks from metal cylinders.

* Dataset information: https://archive.ics.uci.edu/ml/datasets/Connectionist+Bench+(Sonar,+Mines+vs.+Rocks) 
* Dataset Download link: https://archive.ics.uci.edu/ml/machine-learning-databases/undocumented/connectionist-bench/sonar/sonar.all-data

In [84]:
# Require libraries:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sb

# For warning ignorance
import warnings
warnings.filterwarnings('ignore')

In [85]:
# Let's first read dataset:
df = pd.read_csv("sonar_dataset.csv")
df.head()

0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109  \
0  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
1  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
2  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
3  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   
4  0.0286  0.0453  0.0277  0.0174  0.0384  0.0990  0.1201  0.1833  0.2105   

   0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084  \
0  0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
1  0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   
2  0.1264  ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044   
3  0.4459  ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048   
4  0.3039  ...  0.0045  0.0014  0.0038  0.0013  0.0089  0.0057  0.0027   

   0.0090  0.0032  R  
0  0.0052  0.0044  R  
1  0.0095  0.0078  R  
2  0.0040  0.0117  R  
3  0.0107  0.0094  R  
4  0.0051  0.0062  R  

[5 rows x 61 columns]

In [86]:
# We ignore the header because the dataset doesn't has column names, so it will use integer sequence as column names:
df = pd.read_csv("sonar_dataset.csv", header=None)
df.head()

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      51      52      53      54      55      56      57  \
0  0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084   
1  0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
2  0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   
3  0.1264  ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044   
4  0.4459  ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048   

       58      59  60  
0  0.0090  0.0032   R  
1  0.0052  0.0044   R  
2  0.0095  0.0078   R  
3  0.0040  0.0117   R  
4  0.0107  0.0094   R  

[5 rows x 61 columns]

In [16]:
# Let's see the columns names which is numeric integer from 0 to 60:
df.columns

Int64Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
            17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
            34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
            51, 52, 53, 54, 55, 56, 57, 58, 59, 60],
           dtype='int64')

* So here from 0 to 59 is our independent variables and the 60th one is our target variable.

In [9]:
# Let's see the number of rows and columns in the dataset:
df.shape

(208, 61)

In [13]:
# To check n.a. values in the dataset:
df.isnull().sum()

0     0
1     0
2     0
3     0
4     0
     ..
56    0
57    0
58    0
59    0
60    0
Length: 61, dtype: int64

In [23]:
# To check the number of different values (number of classes) in the target variable:
df[60].value_counts()

M    111
R     97
Name: 60, dtype: int64

* So we have 111 values of 'M' and 97 values of 'R'
* We know it's binary classification problem.

In [87]:
# The next step is to create 'X' and 'Y'.
X = df.drop([60], axis = "columns")
y = df[60]
y.sample(6)

205    M
81     R
99     M
39     R
52     R
176    M
Name: 60, dtype: object

In [88]:
# So as we see the 'y' is text data and we need to convert it into integer. So we just use get_dummies method to do One Hot Encodein and then we drop the first dummy:
y = pd.get_dummies(y, drop_first=True)
y.sample(5) # R --> 1 and M --> 0

R
148  0
191  0
75   1
6    1
60   1

In [32]:
# Again if we do value_count(), we will see the convert result:
y.value_counts()

R
0    111
1     97
dtype: int64

In [89]:
# The next step is to divide the data into train and test dataset:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

In [90]:
# To check the some random train samples:
X_train.sample(5)

0       1       2       3       4       5       6       7       8   \
92   0.0260  0.0192  0.0254  0.0061  0.0352  0.0701  0.1263  0.1080  0.1523   
99   0.0201  0.0423  0.0554  0.0783  0.0620  0.0871  0.1201  0.2707  0.1206   
195  0.0129  0.0141  0.0309  0.0375  0.0767  0.0787  0.0662  0.1108  0.1777   
113  0.0283  0.0599  0.0656  0.0229  0.0839  0.1673  0.1154  0.1098  0.1370   
157  0.0201  0.0178  0.0274  0.0232  0.0724  0.0833  0.1232  0.1298  0.2085   

         9   ...      50      51      52      53      54      55      56  \
92   0.1630  ...  0.0132  0.0118  0.0120  0.0051  0.0070  0.0015  0.0035   
99   0.0279  ...  0.0081  0.0191  0.0182  0.0160  0.0290  0.0090  0.0242   
195  0.2245  ...  0.0204  0.0124  0.0093  0.0072  0.0019  0.0027  0.0054   
113  0.1767  ...  0.0109  0.0147  0.0170  0.0158  0.0046  0.0073  0.0054   
157  0.2720  ...  0.0253  0.0131  0.0049  0.0104  0.0102  0.0092  0.0083   

         57      58      59  
92   0.0008  0.0044  0.0077  
99   0.0224  0.0190  0.0096  
195  0.0017  0.0024  0.0029  
113  0.0033  0.0045  0.0079  
157  0.0020  0.0048  0.0036  

[5 rows x 60 columns]

In [91]:
# And also some random test samples:
X_test.sample(5)

0       1       2       3       4       5       6       7       8   \
169  0.0130  0.0120  0.0436  0.0624  0.0428  0.0349  0.0384  0.0446  0.1318   
34   0.0311  0.0491  0.0692  0.0831  0.0079  0.0200  0.0981  0.1016  0.2025   
35   0.0206  0.0132  0.0533  0.0569  0.0647  0.1432  0.1344  0.2041  0.1571   
151  0.0231  0.0315  0.0170  0.0226  0.0410  0.0116  0.0223  0.0805  0.2365   
69   0.0216  0.0215  0.0273  0.0139  0.0357  0.0785  0.0906  0.0908  0.1151   

         9   ...      50      51      52      53      54      55      56  \
169  0.1375  ...  0.0024  0.0084  0.0100  0.0018  0.0035  0.0058  0.0011   
34   0.0767  ...  0.0089  0.0087  0.0032  0.0130  0.0188  0.0101  0.0229   
35   0.1573  ...  0.0307  0.0386  0.0147  0.0018  0.0100  0.0096  0.0077   
151  0.2461  ...  0.0151  0.0125  0.0036  0.0123  0.0043  0.0114  0.0052   
69   0.0973  ...  0.0082  0.0140  0.0044  0.0052  0.0073  0.0021  0.0047   

         57      58      59  
169  0.0009  0.0033  0.0026  
34   0.0182  0.0046  0.0038  
35   0.0180  0.0109  0.0070  
151  0.0091  0.0008  0.0092  
69   0.0024  0.0009  0.0017  

[5 rows x 60 columns]

In [92]:
# The next step is to create Artificial Neural Network (ANN), so for that we need to import tensorflow and keras:
import tensorflow as tf
from tensorflow import keras as kr

In [93]:
# So for ANN model creation we call to sequential function and define the parameters:
model = kr.Sequential([
    kr.layers.Dense(60, input_dim=60, activation='relu'),
    kr.layers.Dense(30, activation='relu'),
    kr.layers.Dense(15, activation='relu'),
    kr.layers.Dense(1, activation='sigmoid')
])

model.compile(
    optimizer='adam',
    loss='binary_crossentropy', 
    metrics=['accuracy']
)

model.fit(X_train, y_train, epochs=100, batch_size=8) # here we use mini-batch gradient descent. after each 8 samples the error will be calculated and then back propagation is done.

Epoch 1/100
20/20 [==============================] - 1s 2ms/step - loss: 0.7073 - accuracy: 0.4487
Epoch 2/100
20/20 [==============================] - 0s 4ms/step - loss: 0.6908 - accuracy: 0.5064
Epoch 3/100
20/20 [==============================] - 0s 3ms/step - loss: 0.6865 - accuracy: 0.5705
Epoch 4/100
20/20 [==============================] - 0s 3ms/step - loss: 0.6800 - accuracy: 0.5962
Epoch 5/100
20/20 [==============================] - 0s 3ms/step - loss: 0.6703 - accuracy: 0.6923
Epoch 6/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6597 - accuracy: 0.7115
Epoch 7/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6484 - accuracy: 0.6731
Epoch 8/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6244 - accuracy: 0.7308
Epoch 9/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5962 - accuracy: 0.7500
Epoch 10/100
20/20 [==============================] - 0s 3ms/step - loss: 0.5725 - accuracy: 0.7051
Epoch 11/

20/20 [==============================] - 0s 5ms/step - loss: 0.0181 - accuracy: 1.0000
Epoch 84/100
20/20 [==============================] - 0s 5ms/step - loss: 0.0190 - accuracy: 1.0000
Epoch 85/100
20/20 [==============================] - 0s 4ms/step - loss: 0.0191 - accuracy: 1.0000
Epoch 86/100
20/20 [==============================] - 0s 5ms/step - loss: 0.0183 - accuracy: 1.0000
Epoch 87/100
20/20 [==============================] - 0s 4ms/step - loss: 0.0165 - accuracy: 1.0000
Epoch 88/100
20/20 [==============================] - 0s 4ms/step - loss: 0.0161 - accuracy: 1.0000
Epoch 89/100
20/20 [==============================] - 0s 4ms/step - loss: 0.0167 - accuracy: 1.0000
Epoch 90/100
20/20 [==============================] - 0s 4ms/step - loss: 0.0130 - accuracy: 1.0000
Epoch 91/100
20/20 [==============================] - 0s 4ms/step - loss: 0.0149 - accuracy: 1.0000
Epoch 92/100
20/20 [==============================] - 0s 4ms/step - loss: 0.0120 - accuracy: 1.0000
Epoch 93/100


* So after 80 epochs we got 100% accuracy.

In [94]:
# So we want to overfit this model. Now the model did really good. let's see the performance on test dataset:
model.evaluate(X_test, y_test)

2/2 [==============================] - 0s 4ms/step - loss: 0.8057 - accuracy: 0.7500


[0.8056643605232239, 0.75]

In [95]:
# To see some prediction, how the model is good:

y_pred = model.predict(X_test).reshape(-1)
print(y_pred[:10])

2/2 [==============================] - 0s 4ms/step
[2.0224654e-05 9.9254948e-01 6.2775713e-01 1.5926962e-04 9.9999934e-01
 9.9907148e-01 1.1053902e-01 9.9999988e-01 4.4174894e-04 1.0000000e+00]


In [96]:
# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

[0. 1. 1. 0. 1. 1. 0. 1. 0. 1.]


In [97]:
# Now to display y_test and compare it with the model prediction:
y_test[:10]

R
186  0
155  0
165  0
200  0
58   1
34   1
151  0
18   1
202  0
62   1

In [98]:
# To see the classification report:
from sklearn.metrics import confusion_matrix , classification_report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.71      0.89      0.79        27
           1       0.83      0.60      0.70        25

    accuracy                           0.75        52
   macro avg       0.77      0.74      0.74        52
weighted avg       0.77      0.75      0.74        52



In [99]:
# So now let's try the model with a dropout layer:
modeld = kr.Sequential([
    kr.layers.Dense(60, input_dim=60, activation='relu'),
    kr.layers.Dropout(0.5), # It will drop the 50% of the neurons.
    kr.layers.Dense(30, activation='relu'),
    kr.layers.Dropout(0.5),
    kr.layers.Dense(15, activation='relu'),
    kr.layers.Dropout(0.5),
    kr.layers.Dense(1, activation='sigmoid')
])

modeld.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

modeld.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 [==============================] - 1s 2ms/step - loss: 0.7503 - accuracy: 0.4808
Epoch 2/100
20/20 [==============================] - 0s 2ms/step - loss: 0.7401 - accuracy: 0.4615
Epoch 3/100
20/20 [==============================] - 0s 3ms/step - loss: 0.7156 - accuracy: 0.4808
Epoch 4/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6983 - accuracy: 0.5256
Epoch 5/100
20/20 [==============================] - 0s 3ms/step - loss: 0.6813 - accuracy: 0.5833
Epoch 6/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6890 - accuracy: 0.5064
Epoch 7/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6950 - accuracy: 0.5064
Epoch 8/100
20/20 [==============================] - 0s 3ms/step - loss: 0.6812 - accuracy: 0.5449
Epoch 9/100
20/20 [==============================] - 0s 3ms/step - loss: 0.6798 - accuracy: 0.5577
Epoch 10/100
20/20 [==============================] - 0s 3ms/step - loss: 0.6674 - accuracy: 0.5705
Epoch 11/

20/20 [==============================] - 0s 4ms/step - loss: 0.2717 - accuracy: 0.8974
Epoch 84/100
20/20 [==============================] - 0s 2ms/step - loss: 0.2801 - accuracy: 0.9038
Epoch 85/100
20/20 [==============================] - 0s 2ms/step - loss: 0.2746 - accuracy: 0.9038
Epoch 86/100
20/20 [==============================] - 0s 2ms/step - loss: 0.3117 - accuracy: 0.8526
Epoch 87/100
20/20 [==============================] - 0s 3ms/step - loss: 0.3194 - accuracy: 0.8526
Epoch 88/100
20/20 [==============================] - 0s 3ms/step - loss: 0.3103 - accuracy: 0.8590
Epoch 89/100
20/20 [==============================] - 0s 3ms/step - loss: 0.2801 - accuracy: 0.8910
Epoch 90/100
20/20 [==============================] - 0s 3ms/step - loss: 0.2897 - accuracy: 0.9231
Epoch 91/100
20/20 [==============================] - 0s 3ms/step - loss: 0.2396 - accuracy: 0.8974
Epoch 92/100
20/20 [==============================] - 0s 3ms/step - loss: 0.3251 - accuracy: 0.8718
Epoch 93/100


In [100]:
# Now we find 88% accuracy on train dataset. The previous one was 100%. We see that the dropout layer reduce the accuracy on
# train dataset, but it's ok. We take care most about test set. The previous test set score were 77%. let's see how many is
# it here right now?
modeld.evaluate(X_test, y_test)

2/2 [==============================] - 0s 4ms/step - loss: 0.4266 - accuracy: 0.7885


[0.4266386032104492, 0.7884615659713745]

* Now as we can see that by using dropout layer we got better performance.

In [102]:
# Now again if we compare the results:
y_pred = modeld.predict(X_test).reshape(-1)
print(y_pred[:10])

y_pred = np.round(y_pred)
print(y_pred[:10])

2/2 [==============================] - 0s 4ms/step
[2.1406219e-04 7.8703457e-01 8.5538918e-01 1.0716150e-02 9.9917203e-01
 9.3518633e-01 3.9348301e-01 9.9965847e-01 2.2632264e-02 9.9980974e-01]
[0. 1. 1. 0. 1. 1. 0. 1. 0. 1.]


In [104]:
y_test[:10]

R
186  0
155  0
165  0
200  0
58   1
34   1
151  0
18   1
202  0
62   1

In [105]:
# To print the classification report from the new model:
from sklearn.metrics import confusion_matrix , classification_report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.77      0.85      0.81        27
           1       0.82      0.72      0.77        25

    accuracy                           0.79        52
   macro avg       0.79      0.79      0.79        52
weighted avg       0.79      0.79      0.79        52



* We see the improvements in classification report.